src/mailroom.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+""" This is the mailroom madness script. It will run from the cmd line and
+ allow you to generate an email or report based on user input."""
+
+import os
+import sys
+import math
+
+
+DONOR_DICT = {'Linus Torvalds': [100, 150, 250],
+              'Jane Goodall': [200, 250, 50],
+              'Bob Ross': [1000, 550, 1]}
+
+
+def main_menu_input():
+    '''Display user prompt for main menu'''
+    user_input = input('''
+    Main Menu:
+    E to generate a thank you email to send to a donor
+    R for a report on the donation history of all donors
+    Type exit at any prompt to exit the program.
+
+    >>>''').upper()
+    menu_input_validation(user_input)
+
+
+def email_donor_input():
+    user_input = input('''
+    Please enter in a donor name.
+    Type list for a list of all donors.
+    Type menu to return to the main menu prompt.
+
+    >>>''').upper()
+    email_input_validation(user_input)
+
+
+def email_donation_input():
+    user_input = int(input('''
+    Please enter the donation amount in numbers only.
+    >>>'''))
+    if type(user_input) == int or type(user_input) == float:
+        return float(user_input)
+    else:
+        email_donation_input()
+
+
+def menu_input_validation(user_input):
+    '''Validate user input from the main menu'''
+    if user_input != 'EXIT' and user_input != 'E' and user_input != 'R':
+        print('\nInvalid input. A valid input is X, E or R.\n')
+        main_menu_input()
+    else:
+        main_menu_route(user_input)
+
+
+def email_input_validation(user_input):
+    '''Validate user input from the email menu'''
+    if user_input == 'LIST':
+        list_donors(DONOR_DICT)
+    elif user_input == 'MENU':
+        main_menu_input()
+    elif user_input == 'EXIT':
+        sys.exit(0)
+    else:
+        donor_db_check(user_input)
+
+
+def main_menu_route(user_input):
+    if user_input == 'X':
+        sys.exit()
+    elif user_input == 'R':
+        report()
+    elif user_input == 'E':
+        email_donor_input()
+
+
+def report():
+    '''Displays donor stats report'''
+    os.system('clear')
+    list_build = donor_stat(DONOR_DICT)
+    dummy_donor = ['Donor Name', '# of Donations', 'Avg $ Amount',
+                   'Total $ Amount']
+    print('-' * 90)
+    print('| {:^30} | {:^15} | {:<16} | {:<16} |'.format(dummy_donor[0],
+          dummy_donor[1], dummy_donor[2], dummy_donor[3]))
+    for i, donor in enumerate(list_build):
+        print('| {:^30} | {:^15} | ${:<15.2f} | ${:<15.2f} |'
+              .format(donor[0], donor[2], donor[1], donor[3]))
```

Comment:
as we said in class, I would *so* refactor the `donor_stat` function to produce tuples in the right order to use tuple unpacking here.  It would be enormously easier to read.